x-pack/filebeat/input/httpjson/request.go

```diff
@@ -37,8 +37,17 @@ func (c *httpClient) do(stdCtx context.Context, trCtx *transformContext, req *ht
 	if err != nil {
 		return nil, fmt.Errorf("failed to execute http client.Do: %w", err)
 	}
+	defer resp.Body.Close()
+
+	// Read the whole resp.Body so we can release the conneciton.
+	// This implementaion is inspired by httputil.DumpResponse
+	resp.Body, err = drainBody(resp.Body)
```

Comment:
Why do we need the `drainBody` function? I have always used something like `io.Copy(ioutil.Discard, resp.Body)` to do this.

Reply:
You mean using a `io.Copy` to copy to a buffer directly?

The main point here is to copy the response body to a in-memory buffer, and use it in the `http.Response` that is returned. I mostly used the same approach as `httputil.DumpResponse` to copy the body into memory and make it available in the `http.Response`. It looks like a neat implementation.

Reply:
Thanks I needed to read some more of the code to see what is happening. I looked at this a bit too quickly.

The point is to drain the response body as soon as we can, but still allow the `resp.Body` to be read later as if this wasn't taken care of already. I agree this is a neat way to do it.